question22/main.js

```diff
@@ -0,0 +1,188 @@
+
+const userDataColumn = {
+  userId: {
+    value : "ID",
+    sort : true
+  },
+  name: {
+    value : "名前",
+    sort : false
+  },
+  gender: {
+    value : "性別",
+    sort : false
+  },
+  age: {
+    value : "年齢",
+    sort : true
+  }
+}
+
+const renderTableElement = (userData)  => {
+  const table = createElementWithClassName("table", "table");
+  const tbody = createElementWithClassName("tbody", "tbody");
+  table.id = "js-table";
+  tbody.id = "js-tbody";
+  wrapper.appendChild(table).appendChild(getTableHeadElement());
+  wrapper.appendChild(table).appendChild(tbody).appendChild(getTableRowFragment(userData));
+  sortButtonClickEvent(userData);
+}
+
+const getTableHeadElement = () => {
+  const thead = createElementWithClassName("thead", "thead");
+  const tr = createElementWithClassName("tr", "tr");
+  const fragmentThElement = document.createDocumentFragment();
+  const userDataColumnKeys = Object.keys(userDataColumn);
+
+  userDataColumnKeys.forEach((key) => {
+    const th = createElementWithClassName("th", "th");
+    const value = userDataColumn[key].value;
+    th.textContent = value; 
+    userDataColumn[key].sort && th.appendChild(createSortButton());
+    fragmentThElement.appendChild(tr).appendChild(th); 
+  });
+
+  thead.appendChild(fragmentThElement);
+  return thead;
+}
+
+const getTableRowFragment = (userData) => {
+  const tableRowfragment = document.createDocumentFragment();
+  for (let i = 0; i < userData.length; i++) {
+    const tr = createElementWithClassName("tr", "tr");
+    tableRowfragment.appendChild(tr).appendChild(getTableDataFragment(userData[i]));
+  }
+  return tableRowfragment;
+}
+
+const getTableDataFragment = (userData) => {
+  const tableDatafragment = document.createDocumentFragment();
+  Object.keys(userDataColumn).forEach((key) => {
+    const td = createElementWithClassName("td", "td");
+    td.textContent = userData[key];
+    tableDatafragment.appendChild(td);
+  });
+  return tableDatafragment;
+};
+
+const createSortButton = () => {
+  const sortButton = createElementWithClassName("button", "sortButton");
+  sortButton.dataset.status = "default";
+  sortButton.id = "js-sortbtn";
+  return sortButton;
+}
+
+const sortButtonClickEvent = () => {
+  const tbody = document.querySelector('tbody');
+  const sortButtons = [...document.querySelectorAll(".sortButton")];
+  const trElement = [...document.querySelectorAll("tbody > tr")];
+  sortButtons.forEach((sortButton)=> {
+    sortButton.addEventListener('click', (e) => {
+      resetSortbuttonStatus(sortButtons, e.target);
+      const clickedCellIndex = e.target.parentElement.cellIndex;
+      const nextStatus = switchSortStatus(e.target);
+      sortButton.dataset.status = nextStatus;
+      const sortedRows = sortFunc(e.target,clickedCellIndex,trElement);
+      while (tbody.firstChild) {
+        tbody.removeChild(tbody.firstChild);
+    }
+      sortedRows.forEach((row) => {
+        tbody.appendChild(row);
+      });
+    });
+  })
+}
```

Comment:
IMO
This function, I feel, has two kinds of jobs.
The function that makes the button click and the function that gives the click function.
Separating them makes the outlook even better.

Reply:
I am sorry, but I am not confident that I fully understand your intent. I tried to separate the process when clicked. Is this intention correct...?
fixed commit : https://github.com/csakyo/homework/pull/37/commits/fa74638321c871b494288d2f22617530db6f5700

Reply:
Thanks for the fix!
I felt that the code here is very good, but it would be better to define it as an object as much as possible because the more arguments it has, the less readable it may become.
[関数の引数が多い場合、Patterns Objectを利用する](https://zenn.dev/chida/articles/ff3e073bce9306#7.%E9%96%A2%E6%95%B0%E3%81%AE%E5%BC%95%E6%95%B0%E3%81%8C%E5%A4%9A%E3%81%84%E5%A0%B4%E5%90%88%E3%80%81patterns-object%E3%82%92%E5%88%A9%E7%94%A8%E3%81%99%E3%82%8B)

```js
const addClickEventToSortButton = () => {
  const sortButtons = [...document.querySelectorAll(".sortButton")];
  const trElement = [...document.querySelectorAll("tbody > tr")];
  sortButtons.forEach((sortButton) => {
    sortButton.addEventListener("click", (e) => {
      clickSortButton({
        target: e.target,
        sortButtons: sortButtons,
        trElement: trElement
      });
    });
  });
};

const clickSortButton = ({ target, sortButtons, trElement }) => {
  const tbody = document.querySelector("tbody");
  resetSortbuttonStatus(sortButtons, target);
  const clickedCellIndex = target.parentElement.cellIndex;
  const nextStatus = switchSortStatus(target);
  target.dataset.status = nextStatus;
  const sortedRows = sortFunc(target, clickedCellIndex, trElement);
  while (tbody.firstChild) {
    tbody.removeChild(tbody.firstChild);
  }
  sortedRows.forEach((row) => {
    tbody.appendChild(row);
  });
};

```
It seems that there is also a way to write with handleEvent.
```js
const addClickEventToSortButton = () => {
  const sortButtons = [...document.querySelectorAll(".sortButton")];
  const trElement = [...document.querySelectorAll("tbody > tr")];
  sortButtons.forEach((sortButton) => {
    sortButton.addEventListener("click", {
      sortButtons: sortButtons,
      trElement: trElement,
      handleEvent: clickSortButton
    });
  });
};

function clickSortButton(event) {
  const tbody = document.querySelector("tbody");
  resetSortbuttonStatus(this.sortButtons, event.target);
  const clickedCellIndex = event.target.parentElement.cellIndex;
  const nextStatus = switchSortStatus(event.target);
  event.target.dataset.status = nextStatus;
  const sortedRows = sortFunc(event.target, clickedCellIndex, this.trElement);
  while (tbody.firstChild) {
    tbody.removeChild(tbody.firstChild);
  }
  sortedRows.forEach((row) => {
    tbody.appendChild(row);
  });
}

```

Reply:
Oh...Amazing!!!!!
It's easy to see when you pass in an object as an argument.
I learned a lot about writing with handleEvent!
Thanks @443n0511 🌈

I think it is smarter to omit the object property name and value since they are the same name ✨
``` js
const addClickEventToSortButton = () => {
   // omit
    sortButton.addEventListener("click", {
      sortButtons,
      trElement,
      handleEvent: clickSortButton
    });
  });
};
```